### ISSUE TITLE
[RFC] Adding a new Provider/Module to Registry via GitHub issues

### ISSUE THREAD

**Issue by roni-frantchi** [opentofu/opentofu#916]
This proposal is meant to complement the selected new [Homebrew-inspired registry](https://github.com/opentofu/opentofu/issues/741.), and as an alternative to https://github.com/opentofu/opentofu/issues/895.  

### Have packages submitted via GitHub [issue forms](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/syntax-for-githubs-form-schema):  
![1](https://github.com/opentofu/opentofu/assets/3658029/5a5cf27f-854b-451c-8bc0-872005912b9b)
![3](https://github.com/opentofu/opentofu/assets/3658029/ed2dd7d5-be71-4020-ae14-fa87db2df6e7)

## Implementation Overview
1. Have packages submitted via GitHub [issue forms](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/syntax-for-githubs-form-schema)
1. Have a GitHub action that upon matched GitHub issue creation/edit will use the submitted information to:
1. Validate the information provided
1. When valid  
    1. Create a new branch where the submitted data is captured in the the proper sharded directory structure/metadata-json
    1. Create a Pull Request for said branch [description will say closes: #issue to auto-close that issue when merged]
    1. Create Pull Request for GPG key submission
1. When invalid  
    1. Will comment back on issue mentioning author asking them to fix what isn’t valid
    1. Upon edit GitHub action will trigger once more

## Pros
- A neat, familiar and secure user interface for proposing modules/provider
- Completely automated whenever possible
- Issues provide submission visibility, allow for conversations to occur in context and on issue when manual admin/core team intervention is required
- No CLI or additional installs required to submit a provider, less maintenance on client-side issues/distribution by core team members
- No GitHub tokens or other means of CLI based authentication required
- Abstracts away any codebase implementation detail like directory sharding logic/code intricacies otherwise required when submitting new provider/module/key directly
- Visible history/timeline for submission/acceptance/deny

## Cons
- Will contribute to GitHub action cost (though probably nothing significant) 
- Since the PRs are generated, git history won't track to a _signed_ commit by submitter
- Forks may cause some loss of audit history as Issues do not transfer with forks

### Reuse the approach for when revising _existing_ packages
The approach described on the PRD is for submitting new Provider or Module, but it can also be extended to say, submitting a GPG key rotation, ask for removal or replacement of faulty version or forcefully bump a new version beyond that cron based version bump - all in the same familiar manner.  

### GitHub CLI
I have also had the notion of being able to do so from GitHub CLI by going:
```
gh issue create -R 'opentofu/registry-stable' -T provider.yml
```

and following form prompt, but it seems like [issue forms aren’t supported yet by CLI](https://github.com/cli/cli/issues/5865), but perhaps would be a good CLI approach patching into the same mechanism later once GitHub supports forms in its CLI.

**Comment by kislerdm** [comment]
Hey Roni! Thanks for a very clear illustration of the proposal.

I'd like to share a couple of minor comment.

> Forks may cause some loss of audit history as Issues do not transfer with forks

AFAIK if the merge commit contains the issue number+the link, it can be back-tracked to the original repo from the fork. 

> Since the PRs are generated, git history won't track to a signed commit by submitter

Can we use submitted GPG key to sign the commit? Alternatively, we could request the submitter to confirm DCO in the gitops style, i.e. as a comment in PR. WDYT?

**Comment by janosdebugs** [comment]
One more note on my side: the issue text can be changed on GitHub by the author, so auditability can be destroyed this way too.

**Comment by kislerdm** [comment]
@janosdebugs Could you please clarify if you meant the "title" by the term "issue text"? Thanks!

IMO such edge case can be handled by using the provided `namespace` and `name` to generate a PR commit's message. WDYT?

**Comment by janosdebugs** [comment]
I think this would need to have some sort of an audit trail who submitted the issue, if they were a member of which organization at the time, etc. (username, org membership, etc can change over time and the user can be deleted, so this info would be lost if we don't record it in the automated commit).

**Comment by Yantrio** [comment]
> I think this would need to have some sort of an audit trail who submitted the issue,

Work similar to this is planned for gpg key upload verification (see RFC [here](https://github.com/opentofu/opentofu/issues/832)). Maybe we could overlap some of this work with a "provider|module uploader check" to ensure the person uploading the provider is who they say they are, and have the rights to upload.

**Comment by roni-frantchi** [comment]
>  One more note on my side: the issue text can be changed on GitHub by the author, so auditability can be destroyed this way too.

1. I think these can be passed along to the generated PR 
2. GitHub issue edits are still tracked and shown as history and actually render diff:
<img width="923" alt="Screenshot 2023-11-23 at 13 44 10" src="https://github.com/opentofu/opentofu/assets/3658029/2225bef8-1c08-4d5c-842d-4e043a4f369e">

<img width="671" alt="Screenshot 2023-11-23 at 13 44 18" src="https://github.com/opentofu/opentofu/assets/3658029/7e536f26-18d2-4813-ba7e-459ed5dd5234">

**Comment by roni-frantchi** [comment]
> AFAIK if the merge commit contains the issue number+the link, it can be back-tracked to the original repo from the fork.

Yes @kislerdm absolutely that is the intention (will revise RFC summary to say so).  
I think the "history loss" is more in kinda forcing the author be the one making the commit + sign it.  

> Can we use submitted GPG key to sign the commit? Alternatively, we could request the submitter to confirm DCO in the gitops style, i.e. as a comment in PR. WDYT?

Yes, though the more I think of it, I really think we should not deal with it and here's why:  

The person making the submission, did not really submit any code.  
They have submitted a form (could have been our own UI), which we process (could have been our own backend instead of GitHub Action), and persist (could a different persistence layer, just happens to be git).  

In other words in my mind we should not attempt to alter history - events are captured as they happened, including any processing and object persistenting - but it's us who own this part really.  

I wouldn't want to sign on a code contribution that's been manipulated by some process - it's ok if you associate me with it- show how I submitted it (link to issue, PR, captured as they were, audit), but I think that's it.

**Comment by kislerdm** [comment]
> I wouldn't want to sign on a code contribution that's been manipulated by some process 

It's a fair point, thanks for sharing. However, in the case of provider/module submission, I'd consider DCO signature an equivalent of signing T&C when creating new provider in the terraform registry. WDYT?

**Comment by roni-frantchi** [comment]
> I'd consider DCO signature an equivalent of signing T&C when creating new provider in the terraform registry. WDYT?

Fair point @kislerdm .  

How about this then? -

<img width="951" alt="image" src="https://github.com/opentofu/opentofu/assets/3658029/ac6f963f-fb78-4c2c-a78d-5d56f6b5fe18">

**Comment by kislerdm** [comment]
@roni-frantchi beautiful! 🎉

**Comment by RLRabinowitz** [comment]
@roni-frantchi I don't think a GHA can create a fork for a user submitting the form. But do we need forks here at all anyway?
I think we can simply create PRs in the registry repository itself
